### PR TITLE
Adding SPY 0.0.3

### PR DESCRIPTION
**Pull request by jfalcou**
Specify library name and version:  **lib/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/spy/0.0.3/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_library(spy_config INTERFACE)
+target_compile_features ( spy_config INTERFACE  cxx_std_17 )
+
+add_executable(example example.cpp)
+target_link_libraries(example ${CONAN_LIBS} PUBLIC spy_config)
+
+# CTest is a testing tool that can be used to test your project.
```
Comment (SSE4) [inline recipes/spy/0.0.3/test_package/CMakeLists.txt:13]: remove dead code

File: recipes/spy/0.0.3/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://jfalcou.github.io/spy/"
+    description = "C++ 17 for constexpr-proof detection and classification of informations about OS, compiler, etc..."
+    topics = ("c++17", "config", "metaprogramming")
+    no_copy_source = True
+    scm = {
+        "type": "git",
+        "url": "auto",
+        "revision": "auto"
+    }
```
Comment (ericLemanissier) [inline recipes/spy/0.0.3/conanfile.py:19]:
```suggestion
```
SCM attribute does not make sense for recipes hosted on CCI, the sources are not on the same repo

Comment (jfalcou) [reply recipes/spy/0.0.3/conanfile.py:19]: dully noted

File: recipes/spy/0.0.3/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import os
+from conans import ConanFile, tools
+
+
+class SpyConan(ConanFile):
+    name = "spy"
+    settings = "os", "compiler", "build_type", "arch"
```
Comment (Minimonium) [inline recipes/spy/0.0.3/conanfile.py:7]: If all you do is just copy a file - you don't need settings.

File: recipes/spy/0.0.3/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+project(PackageTest CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_library(spy_config INTERFACE)
+target_compile_features ( spy_config INTERFACE  cxx_std_17 )
+
+add_executable(example example.cpp)
+target_link_libraries(example ${CONAN_LIBS} PUBLIC spy_config)
```
Comment (uilianries) [inline recipes/spy/0.0.3/test_package/CMakeLists.txt:11]:
```suggestion
add_executable(example example.cpp)
target_link_libraries(example ${CONAN_LIBS})
set_property(TARGET example PROPERTY CXX_STANDARD 17)
```

keep it simple

File: recipes/spy/0.0.3/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        cmake = CMake(self)
+        # Current dir is "test_package/build/<build_id>" and CMakeLists.txt is
+        # in "test_package"
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
+            os.chdir("bin")
+            self.run(".%sexample" % os.sep)
```
Comment (uilianries) [inline recipes/spy/0.0.3/test_package/conanfile.py:18]:
```suggestion
            bin_path = os.path.join("bin", "example")
            self.run(bin_path, run_environment=True)
```

File: recipes/spy/0.0.3/conanfile.py [hunk 6]
```diff
@@ -14,6 +16,16 @@ class SpyConan(ConanFile):
...
 
+    def _supports_cpp17(self):
+        supported_compilers = [("gcc", "7"), ("clang", "5"), ("apple-clang", "10"), ("Visual Studio", "15.7")]
+        compiler = self.settings.compiler
+        version = Version(self.settings.compiler.version)
+        return any(compiler == e[0] and version >= e[1] for e in supported_compilers)
+
+    def configure(self):
+        if not self._supports_cpp17():
+            raise ConanInvalidConfiguration("Absent requires C++17 support")
```
Comment (Minimonium) [inline recipes/spy/0.0.3/conanfile.py:27]: /s/Absent/Spy

File: recipes/spy/0.0.3/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    version = "0.0.3"
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://jfalcou.github.io/spy/"
+    description = "C++ 17 for constexpr-proof detection and classification of informations about OS, compiler, etc..."
+    topics = ("c++17", "config", "metaprogramming")
+    no_copy_source = True
+
+    _source_subfolder = "source_subfolder"
+
```
Comment (uilianries) [inline recipes/spy/0.0.3/conanfile.py:18]:
```suggestion
    def configure(self):
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, "17")
        compiler = self.settings.compiler
        comp_version = self.settings.compiler.version
        compilers = {"Visual Studio": "16", "gcc": "7.3", "clang": "6", "apple-clang": "10.0"}
        min_version = compilers.get(compiler)
        if not min_version:
            self.output.warn("{} recipe lacks information about the {} compiler standard version support.".format(self.name, comp_version))

        elif Version(comp_version) < min_version:
            raise ConanInvalidConfiguration("{} requires a compiler that supports"
                                            " at least C++17. {} {} is not" 
                                            " supported.".format(self.name, compiler, comp_version))
```

C++17 is not supported by all compilers

File: recipes/spy/0.0.3/conanfile.py (hunk 6 above)
Comment (Minimonium) [reply recipes/spy/0.0.3/conanfile.py:27]:
```suggestion
            raise ConanInvalidConfiguration("Spy requires C++17 support")
```

File: recipes/spy/0.0.3/conanfile.py (hunk 7 above)
Comment (Minimonium) [reply recipes/spy/0.0.3/conanfile.py:18]: There is already a compiler version check, was added while you were writing the suggestion. 🙂  We don't really have a unified check now, so I'd like to keep it simple [for a few weeks?] until we release at least some of the cppstd tools.

File: recipes/spy/0.0.3/test_package/CMakeLists.txt (hunk 4 above)
Comment (jfalcou) [reply recipes/spy/0.0.3/test_package/CMakeLists.txt:11]: Well, the use of cxx_std_17 is the proper CMake way to pass standard to target as the latest CMake doc advises.

File: recipes/spy/0.0.3/conanfile.py (hunk 7 above)
Comment (jfalcou) [reply recipes/spy/0.0.3/conanfile.py:18]: This has been added at line 19->27

File: recipes/spy/0.0.3/test_package/CMakeLists.txt (hunk 4 above)
Comment (uilianries) [reply recipes/spy/0.0.3/test_package/CMakeLists.txt:11]: target_compile_features is CMake 3+ Almost all official recipes use this approach.

File: recipes/spy/0.0.3/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import os
+from conans import ConanFile, tools, errors
```
Comment (uilianries) [inline recipes/spy/0.0.3/conanfile.py:2]:
```suggestion
from conans import ConanFile, tools
```

unused import

**Pull request merged**